torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -341,14 +341,9 @@ bool ProcessGroupGloo::SendWork::wait() {
   } catch (...) {
     exception = std::current_exception();
   }
-  // Lock to write completed_ and exception_, and throw if there is an
-  // exception.
-  std::lock_guard<std::mutex> lock(mutex_);
-  completed_ = true;
-  exception_ = exception;
-  if (exception_) {
-    std::rethrow_exception(exception_);
-  }
+
+  // Completes the Work object and throws the exception.
+  finishAndThrow(exception);
```

Comment:
curious, why this cannot be moved to the `catch` block, saving a `std::exception_ptr` var?

Reply:
It probably can be moved inside - just that we still need to mark the work object as completed even if there's no exception. Since this logic sequence (mark complete, set exception, then throw if needed) is pretty common across the ProcessGroup backends, it would be nice to have a helper function that does all this and only grabs the lock once.